express/parsers/apps/espresso/formats/txt.py

```diff
@@ -832,8 +832,90 @@ def dielectric_tensor_generic(self, dat_file: Path) -> Optional[np.ndarray]:
         """
         data = None
         try:
-            data = np.loadtxt(dat_file, dtype=np.dtype([("energy", float), ("eps", (float, 3))]),
-                              converters=lambda x: np.nan_to_num(float(x)))
+            data = np.loadtxt(
+                dat_file,
+                dtype=np.dtype([("energy", float), ("eps", (float, 3))]),
+                converters=lambda x: np.nan_to_num(float(x)),
+            )
         except Exception as e:
             print(e)
         return data
+
+    def parse_hubbard_u(self) -> list:
+        """
+        Extract Hubbard parameters produced by hp.x
+
+        filename: __prefix__.Hubbard_parameters.dat
+
+        Example input content:
+        =-------------------------------------------------------------------------------=
+
+                                        Hubbard U parameters:
+
+            site n.  type  label  spin  new_type  new_label  manifold  Hubbard U (eV)
+                1        1    Co1     1      1         Co1        3d       6.7553
+                2        2    Co2    -1      1         Co1        3d       6.7553
+
+        =-------------------------------------------------------------------------------=
+
+        returns list of following (example) data:
+        {
+            "values": [
+                {
+                "siteNumber": 1,
+                "atomicSpecies": "Co1",
+                "orbitalName": "3d",
+                "value": 6.7553
+                },
+                {
+                "siteNumber": 2,
+                "atomicSpecies": "Co2",
+                "orbitalName": "3d",
+                "value": 6.7553
+                }
+            ]
+        """
+        dat_file = find_file(settings.HP_FILE, self.work_dir)
+        with open(dat_file, "r", encoding="utf-8") as fp:
+            data = fp.read()
+
+        RE_HP_HEADER = (r"\s*({0})\s*({1})\s+({2})\s+({3})\s+({4})\s+({5})\s+({6})\s+({7})" r"\s+({8})\s*").format(
+            "Hubbard U parameters:",
+            "site n.",
+            "type",
+            "label",
+            "spin",
+            "new_type",
+            "new_label",
+            "manifold",
+            "Hubbard U \(eV\)",
+        )
+        RE_HP_DATA = r"\s*{0}\s+{0}\s+{1}\s+{0}\s+{0}\s+{1}\s+{2}\s+{3}".format(
+            GENERAL_REGEX["int_number"],
+            settings.HUBBARD_REGEX["atomicSpecies"],
+            settings.HUBBARD_REGEX["orbitalName"],
+            GENERAL_REGEX["double_number"],
+        )
+        RE_HP_BLOCK = r"{0}({1})+".format(RE_HP_HEADER, RE_HP_DATA)
+
+        hp_block = re.search(RE_HP_BLOCK, data, re.MULTILINE).group()
+        hp_data = re.findall(r"^{0}".format(RE_HP_DATA), hp_block, re.MULTILINE)
+
+        values = []
+
+        for line in hp_data:
```

Comment:
Can we define the regular expression with groups to avoid having to use the for loop?